go/conversation/forms.py

```diff
@@ -54,3 +57,79 @@ class ReplyToMessageForm(forms.Form):
         label='Reply Message',
         required=True,
         widget=forms.Textarea)
+
+
+class MessageDownloadForm(forms.Form):
+
+    CUSTOM_DATE_FORMAT = '%Y/%m/%d'
+
+    format = forms.ChoiceField(
+        required=True, initial='csv',
+        choices=[
+            ('csv', 'csv'), ('json', 'json'),
+        ])
+
+    direction = forms.ChoiceField(
+        required=True, initial='inbound',
+        choices=[
+            ('inbound', 'inbound'), ('outbound', 'outbound'),
+        ])
+
+    date_preset = forms.ChoiceField(
+        required=False, initial='all',
+        choices=[
+            ('all', 'all'), ('1d', '1d'), ('7d', '7d'), ('30d', '30d'),
+        ])
+
+    date_from = forms.DateField(
+        required=False,
+        input_formats=[CUSTOM_DATE_FORMAT])
+
+    date_to = forms.DateField(
+        required=False,
+        input_formats=[CUSTOM_DATE_FORMAT])
+
+    def _parse_date_preset(self, preset):
+        if preset == "all":
+            return None, None
+        days = int(preset[:-1])
+        start_time = (
+            datetime.datetime.utcnow() - datetime.timedelta(days=days))
+        return start_time, None
+
+    def _date_to_utc(self, custom_date):
+        if custom_date is None:
+            return None
+        return datetime.datetime(
+            custom_date.year, custom_date.month, custom_date.day,
+            tzinfo=timezone.utc)
+
+    def _format_custom_date_part(self, date, default):
+        if date is None:
+            return default
+        return date.strftime('%Y%m%d')
+
+    def _format_custom_date_filename(self, start_date, end_date):
+        if start_date is None and end_date is None:
+            return "all"
+        return "-".join([
+            self._format_custom_date_part(start_date, 'until'),
+            self._format_custom_date_part(end_date, 'now'),
+        ])
+
+    def date_range(self):
+        """Return the date range.
+
+        :returns:
+            A tuple of `start_date`, `end_date`, `filename_date`.
+        """
+        date_preset = self.cleaned_data.get('date_preset')
+        if date_preset:
+            start_date, end_date = self._parse_date_preset(date_preset)
+            filename_date = date_preset
+        else:
+            start_date = self._date_to_utc(self.cleaned_data.get('date_from'))
+            end_date = self._date_to_utc(self.cleaned_data.get('date_to'))
+            filename_date = self._format_custom_date_filename(
+                start_date, end_date)
+        return start_date, end_date, filename_date
```

Comment:
If I'm understanding things properly, we would get filenames like `conv1-outbound-1d.csv`, `conv-outbound-19000101-now` and `conv-outbound-until-now.csv`. If the user was to be doing downloads quite often, it is likely they would end up with multiple files with the same name, and it might be difficult for them to figure out which is which. I think it would be useful if we could include the actual dates in the filenames to avoid this (e.g, instead of `conv1-outbound-1d.csv`, we have `conv-outbound-19000101-19000102`).

The `until-` case is a bit more difficult I guess, since we don't know the start date, but I think filenames like `conv-outbound-until-19000101` are still useful.

I don't think the same holds for the `-now` case, since 'now' is relative. I think it would useful if we could use the current date instead of 'now' (e.g. `conv-outbound-until-19000101` if 'now' is `19000101`). It would mean we would be giving the message export api an end time instead of having it use the actual time the export is started as the end time, but I think people might find it confusing to find messages that arrived after the time they requested the download to start to appear in their message download. If you're strongly against this, I'm okay with us keeping it at 'now' though.

Thoughts?